js/reference.js

```diff
@@ -323,16 +324,42 @@ var ERMrest = (function(module) {
-                                // fk is in this table, avoid duplicate and it's not hidden.
-                                if (!(fkName in addedFKs) && fk._table == this._table && !hideFKR(fk)) {
-                                    addedFKs[fkName] = true;
-                                    this._referenceColumns.push(new ReferenceColumn(this, fk.colset.columns[0], fk));
+                                fkName = fk.object.constraint_names[0].join(":");
+                                switch(fk.subject) {
+                                    case module._constraintTypes.FOREIGN_KEY:
+                                        fk = fk.object;
+                                        // fk is in this table, avoid duplicate and it's not hidden.
+                                        if (!(fkName in addedFKs) && fk._table == this._table && !hideFKR(fk)) {
+                                            addedFKs[fkName] = true;
+                                            this._referenceColumns.push(new ReferenceColumn(this, (fk.simple ? fk.colset.columns[0] : null), {"foreignKey":fk}));
+                                        }
+                                        break;
+                                    case module._constraintTypes.KEY:
+                                        fk = fk.object;
+                                        // key is in this table, and avoid duplicate
+                                        if (!(fkName in addedKeys) && fk.table == this._table) {
+                                            addedKeys[fkName] = true;
+                                            // if key is not well formed (has notnull or isHTML), or we're in edit context: add its constituent columns
+                                            if (module._isEntryContext(this._context) || !fk._isWellFormed(this._context)) {
+                                                cols = fk.colset.columns;
+                                                for (j = 0; j < cols.length; j++) {
+                                                    col = cols[j];
+                                                    if (!(col.name in consideredColumns) && !hideColumn(col)) {
+                                                        consideredColumns[col.name] = true;
+                                                        this._referenceColumns.push(new ReferenceColumn(this, cols[j]));
+                                                    }
+                                                }
+                                            } else {
+                                                this._referenceColumns.push(new ReferenceColumn(this, (fk.simple ? fk.colset.columns[0] : null), {"key":fk}));
+                                            }
+                                        }
+                                        break;
+                                    default:
+                                        // unknown behaviour, should not happen
```

Comment:
Have this case throw an error. Ideally there would be error handling where `fk.subject` is set to avoid this from occurring. Better to have this error out than to fail silently.